// @opentripplanner/trip-details/esm/styled.js
import styled from "styled-components";
var BaseButton = styled.button.withConfig({
  displayName: "styled__BaseButton",
  componentId: "sc-6q2ok2-0"
})(["background:transparent;border:0;cursor:pointer;display:inline-block;font-size:14px;font-weight:400;line-height:1.42857143;margin:0;padding:0;text-decoration:none;touch-action:manipulation;user-select:none;vertical-align:middle;white-space:nowrap;"]);
export var CaloriesDescription = styled.span.withConfig({
  displayName: "styled__CaloriesDescription",
  componentId: "sc-6q2ok2-1"
})([""]);
export var CaloriesSummary = styled.span.withConfig({
  displayName: "styled__CaloriesSummary",
  componentId: "sc-6q2ok2-2"
})([""]);
export var CO2Description = styled.span.withConfig({
  displayName: "styled__CO2Description",
  componentId: "sc-6q2ok2-3"
})([""]);
export var CO2Summary = styled.span.withConfig({
  displayName: "styled__CO2Summary",
  componentId: "sc-6q2ok2-4"
})([""]);
export var ExpandButton = styled(BaseButton).withConfig({
  displayName: "styled__ExpandButton",
  componentId: "sc-6q2ok2-5"
})(["color:#00f;font-size:16px;margin-left:6px;margin-top:-2px;"]);
export var Fare = styled.span.withConfig({
  displayName: "styled__Fare",
  componentId: "sc-6q2ok2-6"
})([""]);
export var HideButton = styled(BaseButton).withConfig({
  displayName: "styled__HideButton",
  componentId: "sc-6q2ok2-7"
})(["float:right;"]);
export var TNCFare = styled.span.withConfig({
  displayName: "styled__TNCFare",
  componentId: "sc-6q2ok2-8"
})([""]);
export var TNCFareCompanies = styled.span.withConfig({
  displayName: "styled__TNCFareCompanies",
  componentId: "sc-6q2ok2-9"
})(["text-transform:capitalize;"]);
export var Timing = styled.span.withConfig({
  displayName: "styled__Timing",
  componentId: "sc-6q2ok2-10"
})([""]);
export var TransitFare = styled.details.withConfig({
  displayName: "styled__TransitFare",
  componentId: "sc-6q2ok2-11"
})(["display:inline-block;> span{display:block;padding-left:1.75ch;}"]);
export var TransitFareSingle = styled.span.withConfig({
  displayName: "styled__TransitFareSingle",
  componentId: "sc-6q2ok2-12"
})([""]);
export var FlexSummary = styled.span.withConfig({
  displayName: "styled__FlexSummary",
  componentId: "sc-6q2ok2-13"
})([""]);
export var TripDetail = styled.div.withConfig({
  displayName: "styled__TripDetail",
  componentId: "sc-6q2ok2-14"
})(["margin-top:6px;"]);
export var TripDetailDescription = styled.div.withConfig({
  displayName: "styled__TripDetailDescription",
  componentId: "sc-6q2ok2-15"
})(["background-color:#fff;border:1px solid #888;font-size:12px;margin-top:2px;padding:8px;"]);
export var TripDetailIcon = styled.div.withConfig({
  displayName: "styled__TripDetailIcon",
  componentId: "sc-6q2ok2-16"
})(["float:left;font-size:17px;"]);
export var TripDetails = styled.div.withConfig({
  displayName: "styled__TripDetails",
  componentId: "sc-6q2ok2-17"
})(["background-color:#eee;border-radius:6px;margin-bottom:15px;margin-top:16px;padding:10px 16px;"]);
export var TripDetailsBody = styled.div.withConfig({
  displayName: "styled__TripDetailsBody",
  componentId: "sc-6q2ok2-18"
})([""]);
export var TripDetailsHeader = styled.h2.withConfig({
  displayName: "styled__TripDetailsHeader",
  componentId: "sc-6q2ok2-19"
})(["font-size:18px;font-weight:600;margin:0;"]);
export var TripDetailSummary = styled.div.withConfig({
  displayName: "styled__TripDetailSummary",
  componentId: "sc-6q2ok2-20"
})(["margin-left:28px;padding-top:2px;display:flex;align-items:baseline;"]);
