// @opentripplanner/itinerary-body/esm/stories/itinerary-body-defaults-wrapper.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import coreUtils from "@opentripplanner/core-utils";
import { TriMetLegIcon } from "@opentripplanner/icons";
import PropTypes from "prop-types";
import React, { Component } from "react";
import { action } from "@storybook/addon-actions";
import ItineraryBody from "..";
import DefaultLineColumnContent from "../defaults/line-column-content";
import DefaultPlaceName from "../defaults/place-name";
import DefaultRouteDescription from "../defaults/route-description";
import DefaultTransitLegSummary from "../defaults/transit-leg-summary";
import { StyledItineraryBody } from "../demos";
import OtpRRStyledItineraryBody from "../otp-react-redux/itinerary-body";

var config = require("../__mocks__/config.json");

var ItineraryBodyDefaultsWrapper = /*#__PURE__*/function (_Component) {
  _inherits(ItineraryBodyDefaultsWrapper, _Component);

  var _super = _createSuper(ItineraryBodyDefaultsWrapper);

  function ItineraryBodyDefaultsWrapper() {
    var _this;

    _classCallCheck(this, ItineraryBodyDefaultsWrapper);

    _this = _super.call(this);

    _this.setLegDiagram = function (leg) {
      _this.setState({
        diagramVisible: leg
      });
    };

    _this.state = {};
    return _this;
  }

  _createClass(ItineraryBodyDefaultsWrapper, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          itinerary = _this$props.itinerary,
          LegIcon = _this$props.LegIcon,
          LineColumnContent = _this$props.LineColumnContent,
          PlaceName = _this$props.PlaceName,
          RouteDescription = _this$props.RouteDescription,
          showAgencyInfo = _this$props.showAgencyInfo,
          showLegIcon = _this$props.showLegIcon,
          showMapButtonColumn = _this$props.showMapButtonColumn,
          showRouteFares = _this$props.showRouteFares,
          showViewTripButton = _this$props.showViewTripButton,
          styledItinerary = _this$props.styledItinerary,
          TimeColumnContent = _this$props.TimeColumnContent,
          toRouteAbbreviation = _this$props.toRouteAbbreviation,
          TransitLegSubheader = _this$props.TransitLegSubheader,
          TransitLegSummary = _this$props.TransitLegSummary,
          AlertToggleIcon = _this$props.AlertToggleIcon,
          AlertBodyIcon = _this$props.AlertBodyIcon;
      var diagramVisible = this.state.diagramVisible;
      var ItineraryBodyComponent;

      switch (styledItinerary) {
        case "pink-legs":
          ItineraryBodyComponent = StyledItineraryBody;
          break;

        case "otp-rr":
          ItineraryBodyComponent = OtpRRStyledItineraryBody;
          break;

        default:
          ItineraryBodyComponent = ItineraryBody;
      }

      return /*#__PURE__*/React.createElement(ItineraryBodyComponent, {
        config: config,
        diagramVisible: diagramVisible,
        frameLeg: action("frameLeg"),
        itinerary: itinerary,
        LegIcon: LegIcon,
        LineColumnContent: LineColumnContent || DefaultLineColumnContent,
        mapillaryKey: "fake key, but ok because the api response is also fake",
        PlaceName: PlaceName || DefaultPlaceName,
        RouteDescription: RouteDescription || DefaultRouteDescription,
        routingType: "ITINERARY",
        setActiveLeg: action("setActiveLeg"),
        setLegDiagram: this.setLegDiagram,
        setViewedTrip: action("setViewedTrip"),
        showAgencyInfo: showAgencyInfo,
        showElevationProfile: true,
        showLegIcon: showLegIcon,
        showMapButtonColumn: showMapButtonColumn,
        showRouteFares: showRouteFares,
        showViewTripButton: showViewTripButton,
        TimeColumnContent: TimeColumnContent,
        toRouteAbbreviation: toRouteAbbreviation,
        TransitLegSubheader: TransitLegSubheader,
        TransitLegSummary: TransitLegSummary || DefaultTransitLegSummary,
        AlertToggleIcon: AlertToggleIcon,
        AlertBodyIcon: AlertBodyIcon
      });
    }
  }]);

  return ItineraryBodyDefaultsWrapper;
}(Component);

export { ItineraryBodyDefaultsWrapper as default };
ItineraryBodyDefaultsWrapper.propTypes = {
  itinerary: coreUtils.types.itineraryType.isRequired,
  LegIcon: PropTypes.elementType,
  LineColumnContent: PropTypes.elementType,
  PlaceName: PropTypes.elementType,
  RouteDescription: PropTypes.elementType,
  showAgencyInfo: PropTypes.bool,
  showLegIcon: PropTypes.bool,
  showMapButtonColumn: PropTypes.bool,
  showRouteFares: PropTypes.bool,
  showViewTripButton: PropTypes.bool,
  styledItinerary: PropTypes.string,
  TimeColumnContent: PropTypes.elementType,
  toRouteAbbreviation: PropTypes.func,
  TransitLegSubheader: PropTypes.elementType,
  TransitLegSummary: PropTypes.elementType,
  AlertToggleIcon: PropTypes.elementType,
  AlertBodyIcon: PropTypes.elementType
};
ItineraryBodyDefaultsWrapper.defaultProps = {
  LegIcon: TriMetLegIcon,
  LineColumnContent: undefined,
  PlaceName: undefined,
  RouteDescription: undefined,
  showAgencyInfo: false,
  showLegIcon: false,
  showMapButtonColumn: true,
  showRouteFares: false,
  showViewTripButton: false,
  styledItinerary: null,
  TimeColumnContent: undefined,
  toRouteAbbreviation: function toRouteAbbreviation(r) {
    return r.toString().substr(0, 2);
  },
  TransitLegSubheader: undefined,
  TransitLegSummary: undefined,
  AlertToggleIcon: undefined,
  AlertBodyIcon: undefined
};
