// @opentripplanner/itinerary-body/esm/AccessLegBody/access-leg-steps.js
import coreUtils from "@opentripplanner/core-utils";
import { DirectionIcon } from "@opentripplanner/icons";
import React from "react";
import PropTypes from "prop-types";
import * as Styled from "../styled";
import MapillaryButton from "./mapillary-button";
export default function AccessLegSteps(_ref) {
  var steps = _ref.steps,
      mapillaryCallback = _ref.mapillaryCallback,
      mapillaryKey = _ref.mapillaryKey;
  return /*#__PURE__*/React.createElement(Styled.Steps, null, steps.map(function (step, k) {
    return /*#__PURE__*/React.createElement(Styled.StepRow, {
      key: k
    }, /*#__PURE__*/React.createElement(Styled.StepIconContainer, null, /*#__PURE__*/React.createElement(DirectionIcon, {
      relativeDirection: step.relativeDirection
    })), /*#__PURE__*/React.createElement(Styled.StepDescriptionContainer, null, coreUtils.itinerary.getStepDirection(step), /*#__PURE__*/React.createElement("span", null, step.relativeDirection === "ELEVATOR" ? " to " : " on "), /*#__PURE__*/React.createElement(Styled.StepStreetName, null, coreUtils.itinerary.getStepStreetName(step)), /*#__PURE__*/React.createElement(MapillaryButton, {
      clickCallback: mapillaryCallback,
      coords: step,
      mapillaryKey: mapillaryKey,
      padLeft: true
    })));
  }));
}
AccessLegSteps.propTypes = {
  steps: coreUtils.types.stepsType.isRequired,
  mapillaryCallback: PropTypes.func,
  mapillaryKey: PropTypes.string
};
AccessLegSteps.defaultProps = {
  mapillaryCallback: null,
  mapillaryKey: null
};
