// @opentripplanner/itinerary-body/esm/otp-react-redux/transit-leg-subheader.js
import coreUtils from "@opentripplanner/core-utils";
import PropTypes from "prop-types";
import React from "react";
import * as Styled from "../styled";
import ViewStopButton from "./view-stop-button";
export default function TransitLegSubheader(_ref) {
  var languageConfig = _ref.languageConfig,
      leg = _ref.leg,
      onStopClick = _ref.onStopClick;
  var from = leg.from;
  var isFlex = coreUtils.itinerary.isFlex(leg);
  var buttonText = languageConfig.stopViewer || "Stop Viewer";
  return /*#__PURE__*/React.createElement(Styled.PlaceSubheader, null, /*#__PURE__*/React.createElement("span", null, "Stop ID ", from.stopCode || from.stopId.split(":")[1]), !isFlex && /*#__PURE__*/React.createElement(ViewStopButton, {
    onStopClick: onStopClick,
    stopId: from.stopId,
    text: buttonText
  }));
}
TransitLegSubheader.propTypes = {
  languageConfig: coreUtils.types.languageConfigType.isRequired,
  leg: coreUtils.types.legType.isRequired,
  onStopClick: PropTypes.func.isRequired
};
