// @opentripplanner/itinerary-body/esm/styled.js
import { Map } from "@opentripplanner/icons";
import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import { CaretDown } from "@styled-icons/fa-solid/CaretDown";
import { CaretUp } from "@styled-icons/fa-solid/CaretUp";
import { ExclamationTriangle } from "@styled-icons/fa-solid/ExclamationTriangle";
import { toModeBorder, toModeColor, toSafeRouteColor } from "./util"; // ////////////////////////////////////////////////
// ///////////// Generic components ///////////////
// ////////////////////////////////////////////////

export var ClearButton = styled.button.withConfig({
  displayName: "styled__ClearButton",
  componentId: "sc-6hmr57-0"
})(["background:transparent;color:inherit;border:0;text-align:inherit;text-decoration:none;&:focus{background-color:", ";outline:0;}&:hover{background-color:", ";}&:active{background-color:", ";}"], function (props) {
  return props.theme.tertiaryColor;
}, function (props) {
  return props.theme.hoverColor;
}, function (props) {
  return props.theme.activeColor;
});
/*
  This is needed to give the offset border look to stacked place rows
  Since the value we have access to is "interlineWithPreviousLeg" then we
  have to show/hide the top border of the div and apply a small offset
*/

export var LightBorderDiv = styled.div.withConfig({
  displayName: "styled__LightBorderDiv",
  componentId: "sc-6hmr57-1"
})(["border-top-style:solid;border-top-width:", ";border-top-color:", ";padding-top:", ";padding-bottom:", ";transform:", ";"], function (props) {
  return props.hideBorder === "true" ? "0" : "2px";
}, function (props) {
  return props.theme.borderColor;
}, function (props) {
  return props.hideBorder === "true" ? "0" : "10px";
}, function (props) {
  return props.hideBorder === "true" ? "10px" : "0";
}, function (props) {
  return props.hideBorder === "true" ? "" : "translateY(-12px)";
});
export var TransparentButton = styled.button.withConfig({
  displayName: "styled__TransparentButton",
  componentId: "sc-6hmr57-2"
})(["background:transparent;border:0;color:inherit;cursor:pointer;font-size:inherit;text-decoration:none;&:focus{}"]);
export var AnchorButton = styled.a.withConfig({
  displayName: "styled__AnchorButton",
  componentId: "sc-6hmr57-3"
})(["background-color:#fff;background-image:none;border-radius:4px;border:1px solid #ccc;color:#333;cursor:pointer;display:inline-block;font-size:14px;font-weight:400;padding:4px 6px;text-align:center;text-decoration:none;touch-action:manipulation;user-select:none;white-space:nowrap;&:hover{color:#333;background-color:#e6e6e6;border-color:#adadad;}&:active{color:#333;background-color:#e6e6e6;background-image:none;border-color:#adadad;outline:0;box-shadow:inset 0 3px 5px rgba(0,0,0,0.125);}&:focus{color:#333;background-color:#e6e6e6;border-color:#8c8c8c;}&:active:hover{color:#333;background-color:#d4d4d4;border-color:#8c8c8c;}"]);
export var LinkButton = styled(TransparentButton).withConfig({
  displayName: "styled__LinkButton",
  componentId: "sc-6hmr57-4"
})(["color:#008;cursor:pointer;margin-left:5px;&:hover{text-decoration:underline;}"]);
export var ViewerButton = styled(LinkButton).withConfig({
  displayName: "styled__ViewerButton",
  componentId: "sc-6hmr57-5"
})(["padding-left:0px;&:before{content:\"|\";color:black;margin-right:5px;}"]); // ////////////////////////////////////////////////
// /////////////// App components /////////////////
// ////////////////////////////////////////////////
// TODO: Can we turn this into a more abstract element to inherit from for other badges?

export var AccessBadge = styled.div.attrs(function (props) {
  return {
    "aria-label": "Travel by ".concat(props.mode)
  };
}).withConfig({
  displayName: "styled__AccessBadge",
  componentId: "sc-6hmr57-6"
})(["color:black;background-color:", ";border:2px solid #bbb;text-align:center;width:25px;height:25px;font-size:1.2em;border-radius:50%;display:flex;align-items:center;justify-content:center;padding-left:1px;"], function (props) {
  return toModeColor(props.mode, props.routeColor);
});
export var AgencyInfo = styled.div.withConfig({
  displayName: "styled__AgencyInfo",
  componentId: "sc-6hmr57-7"
})(["margin-top:5px;a{color:#337ab7;text-decoration:none;}a:hover{text-decoration:underline;}img{margin-left:5px;}"]);
export var CallAheadWarning = styled.div.withConfig({
  displayName: "styled__CallAheadWarning",
  componentId: "sc-6hmr57-8"
})(["color:#b22727;margin-top:5px;"]);
export var BookLaterContainer = styled.div.withConfig({
  displayName: "styled__BookLaterContainer",
  componentId: "sc-6hmr57-9"
})(["bottom:0;left:110px;position:absolute;right:0;top:0;"]);
export var BookLaterInnerContainer = styled.div.withConfig({
  displayName: "styled__BookLaterInnerContainer",
  componentId: "sc-6hmr57-10"
})(["background-color:#fcf9d3;display:table;height:100%;width:100%;"]);
export var BookLaterPointer = styled.div.withConfig({
  displayName: "styled__BookLaterPointer",
  componentId: "sc-6hmr57-11"
})(["border-bottom:16px solid transparent;border-right:16px solid #fcf9d3;border-top:16px solid transparent;height:0;left:94px;position:absolute;top:0;width:0;"]);
export var BookLaterText = styled.div.withConfig({
  displayName: "styled__BookLaterText",
  componentId: "sc-6hmr57-12"
})(["color:#444;display:table-cell;font-style:italic;line-height:0.95;padding:0px 2px;vertical-align:middle;"]);
export var BookTNCRideButton = styled(AnchorButton).withConfig({
  displayName: "styled__BookTNCRideButton",
  componentId: "sc-6hmr57-13"
})([""]);
export var BookTNCRideButtonContainer = styled.div.withConfig({
  displayName: "styled__BookTNCRideButtonContainer",
  componentId: "sc-6hmr57-14"
})(["height:32px;margin-bottom:10px;margin-top:10px;position:relative;"]);
export var TNCTravelTime = styled.div.withConfig({
  displayName: "styled__TNCTravelTime",
  componentId: "sc-6hmr57-15"
})([""]);
export var TNCCost = styled.div.withConfig({
  displayName: "styled__TNCCost",
  componentId: "sc-6hmr57-16"
})([""]);
export var CaretToggle = function CaretToggle(_ref) {
  var expanded = _ref.expanded;
  return expanded ? /*#__PURE__*/React.createElement(CaretUp, {
    size: 15
  }) : /*#__PURE__*/React.createElement(CaretDown, {
    size: 15
  });
};
CaretToggle.propTypes = {
  expanded: PropTypes.bool.isRequired
};
export var Destination = styled.div.withConfig({
  displayName: "styled__Destination",
  componentId: "sc-6hmr57-17"
})(["text-align:center;"]);
export var DetailsColumn = styled(LightBorderDiv).withConfig({
  displayName: "styled__DetailsColumn",
  componentId: "sc-6hmr57-18"
})(["flex:2 2 auto;"]);
export var InnerLine = styled.div.withConfig({
  displayName: "styled__InnerLine",
  componentId: "sc-6hmr57-19"
})(["border-left:", ";height:100%;width:0;position:absolute;left:50%;transform:translateX(-50%);"], function (props) {
  return toModeBorder(props.mode, props.routeColor);
});
export var InterlineDot = styled.div.withConfig({
  displayName: "styled__InterlineDot",
  componentId: "sc-6hmr57-20"
})(["color:#fff;flex:0 0 15px;margin-left:-33px;margin-right:18px;position:relative;z-index:30;"]);
export var InterlineName = styled.div.withConfig({
  displayName: "styled__InterlineName",
  componentId: "sc-6hmr57-21"
})([""]);
export var IntermediateStops = styled.div.withConfig({
  displayName: "styled__IntermediateStops",
  componentId: "sc-6hmr57-22"
})(["display:block;font-size:13px;"]);
export var ItineraryBody = styled.div.withConfig({
  displayName: "styled__ItineraryBody",
  componentId: "sc-6hmr57-23"
})([""]);
export var LegBody = styled.div.withConfig({
  displayName: "styled__LegBody",
  componentId: "sc-6hmr57-24"
})(["color:#676767;font-size:13px;padding:12px 0 18px 4px;"]);
export var LegClickable = styled(TransparentButton).withConfig({
  displayName: "styled__LegClickable",
  componentId: "sc-6hmr57-25"
})(["cursor:pointer;display:table;padding:0;text-align:center;line-height:31px;"]);
export var LegDescription = styled.div.withConfig({
  displayName: "styled__LegDescription",
  componentId: "sc-6hmr57-26"
})(["display:table;> div{display:table-cell;}"]);
export var LegDescriptionHeadsignPrefix = styled.span.withConfig({
  displayName: "styled__LegDescriptionHeadsignPrefix",
  componentId: "sc-6hmr57-27"
})(["font-weight:200;"]);
export var LegDescriptionRouteLongName = styled.div.withConfig({
  displayName: "styled__LegDescriptionRouteLongName",
  componentId: "sc-6hmr57-28"
})(["font-size:13px;font-weight:500;line-height:16px;"]);
export var LegDescriptionRouteShortName = styled.div.withConfig({
  displayName: "styled__LegDescriptionRouteShortName",
  componentId: "sc-6hmr57-29"
})(["font-weight:800;margin-right:6px;"]);
export var LegDescriptionForTransit = styled(LegDescription).withConfig({
  displayName: "styled__LegDescriptionForTransit",
  componentId: "sc-6hmr57-30"
})(["color:#807373;margin-top:5px;"]);
export var LegIconContainer = styled.div.withConfig({
  displayName: "styled__LegIconContainer",
  componentId: "sc-6hmr57-31"
})(["height:24px;width:24px;float:left;margin-right:6px;"]);
export var LegLine = styled.div.withConfig({
  displayName: "styled__LegLine",
  componentId: "sc-6hmr57-32"
})(["position:relative;left:50%;transform:translateX(-50%);height:100%;"]);
export var LineBadgeContainer = styled.div.withConfig({
  displayName: "styled__LineBadgeContainer",
  componentId: "sc-6hmr57-33"
})(["width:30px;height:30px;border-radius:50%;position:absolute;left:50%;top:0;transform:translate(-51%,-10%);"]);
export var LineColumn = styled.div.withConfig({
  displayName: "styled__LineColumn",
  componentId: "sc-6hmr57-34"
})(["flex:0 0 50px;padding-right:5px;"]);
export var PlaceRowWrapper = styled.div.withConfig({
  displayName: "styled__PlaceRowWrapper",
  componentId: "sc-6hmr57-35"
})(["max-width:500px;display:flex;flex-flow:row;"]);
export var PreviewContainer = styled.div.withConfig({
  displayName: "styled__PreviewContainer",
  componentId: "sc-6hmr57-36"
})(["background-color:", ";border-color:", ";border-radius:5px;border-style:solid;border-width:1px;display:inline-block;font-style:normal;margin:0 4px;position:relative;text-align:center;text-decoration:none;width:75%;&:hover{border-color:#d1d5da;background-color:#f6f8fa;}"], function (props) {
  return props.active && "#eee";
}, function (props) {
  return props.active ? "#d1d5da" : "#fff";
});
export var TimeColumn = styled.div.withConfig({
  displayName: "styled__TimeColumn",
  componentId: "sc-6hmr57-37"
})(["flex:0 0 60px;padding-right:5px;font-size:0.9em;"]);
export var MapButton = styled(LinkButton).withConfig({
  displayName: "styled__MapButton",
  componentId: "sc-6hmr57-38"
})(["padding:3px 10px 3px 10px;border:0;margin-top:-15px;width:35px;height:35px;&:hover{cursor:pointer;}"]);
export var MapButtonColumn = styled(LightBorderDiv).withConfig({
  displayName: "styled__MapButtonColumn",
  componentId: "sc-6hmr57-39"
})(["flex:0 0 25px;"]);
export var MapIcon = styled(Map).attrs(function (props) {
  return {
    fill: props.theme.secondaryColor,
    width: 15,
    height: 15,
    role: "img"
  };
}).withConfig({
  displayName: "styled__MapIcon",
  componentId: "sc-6hmr57-40"
})([""]);
export var PlaceDetails = styled.div.withConfig({
  displayName: "styled__PlaceDetails",
  componentId: "sc-6hmr57-41"
})([""]);
export var PlaceHeader = styled.div.withConfig({
  displayName: "styled__PlaceHeader",
  componentId: "sc-6hmr57-42"
})(["display:flex;font-size:1.2em;"]);
export var PlaceName = styled.div.withConfig({
  displayName: "styled__PlaceName",
  componentId: "sc-6hmr57-43"
})(["font-weight:bold;height:1.2em;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;flex:1 1 auto;"]);
export var PlaceSubheader = styled.div.withConfig({
  displayName: "styled__PlaceSubheader",
  componentId: "sc-6hmr57-44"
})(["color:#807373;font-weight:300;padding-left:4px;padding-top:1px;"]);
export var PreviewDiagram = styled(TransparentButton).withConfig({
  displayName: "styled__PreviewDiagram",
  componentId: "sc-6hmr57-45"
})(["padding:2px;width:100%;"]);
export var PreviewDiagramElevationChange = styled.span.withConfig({
  displayName: "styled__PreviewDiagramElevationChange",
  componentId: "sc-6hmr57-46"
})(["font-size:xx-small;"]);
export var PreviewDiagramElevationGain = styled(PreviewDiagramElevationChange).withConfig({
  displayName: "styled__PreviewDiagramElevationGain",
  componentId: "sc-6hmr57-47"
})(["color:#e60000;"]);
export var PreviewDiagramElevationLoss = styled(PreviewDiagramElevationChange).withConfig({
  displayName: "styled__PreviewDiagramElevationLoss",
  componentId: "sc-6hmr57-48"
})(["color:green;"]);
export var PreviewDiagramTitle = styled.div.withConfig({
  displayName: "styled__PreviewDiagramTitle",
  componentId: "sc-6hmr57-49"
})(["font-size:small;"]);
export var RouteBadge = styled.div.withConfig({
  displayName: "styled__RouteBadge",
  componentId: "sc-6hmr57-50"
})(["text-align:center;min-width:30px;min-height:30px;font-size:1.2em;background-color:", ";color:white;border-radius:50%;display:flex;align-items:center;justify-content:center;padding-left:1px;border:1px solid ", ";user-select:none;cursor:default;"], function (props) {
  return toSafeRouteColor(props.routeColor) || props.theme.mainColor;
}, function (props) {
  return props.theme.badgeBorderColor;
});
export var SROnly = styled.span.withConfig({
  displayName: "styled__SROnly",
  componentId: "sc-6hmr57-51"
})(["position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);border:0;"]);
export var SRHidden = styled.span.attrs({
  "aria-hidden": true
}).withConfig({
  displayName: "styled__SRHidden",
  componentId: "sc-6hmr57-52"
})([""]);
export var Steps = styled.div.withConfig({
  displayName: "styled__Steps",
  componentId: "sc-6hmr57-53"
})(["display:block;"]);
export var StepDescriptionContainer = styled.div.withConfig({
  displayName: "styled__StepDescriptionContainer",
  componentId: "sc-6hmr57-54"
})(["margin-left:24px;line-height:1.25em;padding-top:1px;"]);
export var StepsHeader = styled(TransparentButton).withConfig({
  displayName: "styled__StepsHeader",
  componentId: "sc-6hmr57-55"
})(["color:#676767;display:inline-block;font-size:13px;font-style:normal;margin-top:10px;vertical-align:bottom;"]);
export var StepIconContainer = styled.div.withConfig({
  displayName: "styled__StepIconContainer",
  componentId: "sc-6hmr57-56"
})(["fill:#676767;float:left;height:16px;width:16px;"]);
export var StepRow = styled.div.withConfig({
  displayName: "styled__StepRow",
  componentId: "sc-6hmr57-57"
})(["font-size:13px;margin-top:8px;color:#676767;font-style:normal;"]);
export var StepStreetName = styled.span.withConfig({
  displayName: "styled__StepStreetName",
  componentId: "sc-6hmr57-58"
})(["font-weight:500;"]);
export var StopIdSpan = styled.span.withConfig({
  displayName: "styled__StopIdSpan",
  componentId: "sc-6hmr57-59"
})(["font-weight:200;font-size:0.9em;margin-left:10px;"]);
export var StopMarker = styled.div.withConfig({
  displayName: "styled__StopMarker",
  componentId: "sc-6hmr57-60"
})(["float:left;margin-left:-36px;color:#fff;"]);
export var StopName = styled.div.withConfig({
  displayName: "styled__StopName",
  componentId: "sc-6hmr57-61"
})(["color:#676767;margin-top:3px;"]);
export var StopRow = styled.div.withConfig({
  displayName: "styled__StopRow",
  componentId: "sc-6hmr57-62"
})(["z-index:30;position:relative;"]);
export var TransitAlert = styled.a.withConfig({
  displayName: "styled__TransitAlert",
  componentId: "sc-6hmr57-63"
})(["background-color:#eee;border-radius:4px;color:#000;display:block;margin-top:5px;padding:8px;text-decoration:none;"]);
export var TransitAlertBody = styled.div.withConfig({
  displayName: "styled__TransitAlertBody",
  componentId: "sc-6hmr57-64"
})(["font-size:12px;margin-left:30px;white-space:pre-wrap;"]);
export var TransitAlertEffectiveDate = styled.div.withConfig({
  displayName: "styled__TransitAlertEffectiveDate",
  componentId: "sc-6hmr57-65"
})(["margin-top:5px;margin-left:30px;font-size:12px;font-style:italic;"]);
export var TransitAlertHeader = styled.div.withConfig({
  displayName: "styled__TransitAlertHeader",
  componentId: "sc-6hmr57-66"
})(["font-size:14px;margin-left:30px;font-weight:600;"]);
export var TransitAlertIconContainer = styled.div.withConfig({
  displayName: "styled__TransitAlertIconContainer",
  componentId: "sc-6hmr57-67"
})(["float:left;font-size:18px;"]);
export var TransitAlerts = styled.div.withConfig({
  displayName: "styled__TransitAlerts",
  componentId: "sc-6hmr57-68"
})(["display:block;margin-top:3px;"]);
export var TransitAlertToggle = styled(TransparentButton).withConfig({
  displayName: "styled__TransitAlertToggle",
  componentId: "sc-6hmr57-69"
})(["color:#d14727;cursor:pointer;display:inline-block;font-weight:400;margin-top:8px;padding:0;"]);
export var TransitLegDetails = styled.div.withConfig({
  displayName: "styled__TransitLegDetails",
  componentId: "sc-6hmr57-70"
})(["margin-top:5px;"]);
export var TransitLegDetailsHeader = styled.div.withConfig({
  displayName: "styled__TransitLegDetailsHeader",
  componentId: "sc-6hmr57-71"
})(["color:#676767;"]);
export var TransitLegExpandedBody = styled.div.withConfig({
  displayName: "styled__TransitLegExpandedBody",
  componentId: "sc-6hmr57-72"
})(["font-size:14px;"]);
export var TransitLegFare = styled.div.withConfig({
  displayName: "styled__TransitLegFare",
  componentId: "sc-6hmr57-73"
})([""]);
export var TransitLegSummary = styled(TransparentButton).withConfig({
  displayName: "styled__TransitLegSummary",
  componentId: "sc-6hmr57-74"
})(["padding:0;"]);
export var DefaultAlertToggleIcon = styled(ExclamationTriangle).attrs({
  size: 15
}).withConfig({
  displayName: "styled__DefaultAlertToggleIcon",
  componentId: "sc-6hmr57-75"
})([""]);
export var DefaultAlertBodyIcon = styled(ExclamationTriangle).attrs({
  size: 18
}).withConfig({
  displayName: "styled__DefaultAlertBodyIcon",
  componentId: "sc-6hmr57-76"
})([""]);
