// @opentripplanner/printable-itinerary/esm/accessibility-annotation.js
import coreUtils from "@opentripplanner/core-utils";
import { AccessibilityRating } from "@opentripplanner/itinerary-body";
import React from "react";
import PropTypes from "prop-types";
import * as S from "./styled";

var AccessibilityAnnotation = function AccessibilityAnnotation(_ref) {
  var accessibilityScoreGradationMap = _ref.accessibilityScoreGradationMap,
      grayscale = _ref.grayscale,
      leg = _ref.leg,
      LegIcon = _ref.LegIcon;
  return /*#__PURE__*/React.createElement(S.LegAnnotation, null, /*#__PURE__*/React.createElement(S.ModeIcon, null, /*#__PURE__*/React.createElement(LegIcon, {
    leg: leg
  })), leg.accessibilityScore && /*#__PURE__*/React.createElement(AccessibilityRating, {
    gradationMap: accessibilityScoreGradationMap,
    grayscale: grayscale,
    score: leg.accessibilityScore
  }));
};

AccessibilityAnnotation.propTypes = {
  accessibilityScoreGradationMap: PropTypes.shape({
    color: PropTypes.string,
    text: PropTypes.string,
    icon: PropTypes.element
  }),
  grayscale: PropTypes.bool,
  leg: coreUtils.types.legType.isRequired,
  LegIcon: PropTypes.elementType.isRequired
};
AccessibilityAnnotation.defaultProps = {
  accessibilityScoreGradationMap: undefined,
  grayscale: false
};
export default AccessibilityAnnotation;
