// @opentripplanner/printable-itinerary/esm/styled.js
import styled from "styled-components";
export var Leg = styled.div.withConfig({
  displayName: "styled__Leg",
  componentId: "sc-1w8qzv8-0"
})(["margin-bottom:10px;border-top:1px solid grey;padding-top:18px;display:flex;"]);
export var CollapsedTop = styled(Leg).withConfig({
  displayName: "styled__CollapsedTop",
  componentId: "sc-1w8qzv8-1"
})(["border-top:none;padding-top:0;"]);
export var LegBody = styled.div.withConfig({
  displayName: "styled__LegBody",
  componentId: "sc-1w8qzv8-2"
})(["margin-left:10px;"]);
export var LegDetail = styled.div.withConfig({
  displayName: "styled__LegDetail",
  componentId: "sc-1w8qzv8-3"
})(["font-size:14px;margin-top:3px;"]);
export var LegDetails = styled.div.withConfig({
  displayName: "styled__LegDetails",
  componentId: "sc-1w8qzv8-4"
})(["margin-top:5px;"]);
export var LegHeader = styled.div.withConfig({
  displayName: "styled__LegHeader",
  componentId: "sc-1w8qzv8-5"
})(["font-size:18px;"]);
export var LegAnnotation = styled.div.withConfig({
  displayName: "styled__LegAnnotation",
  componentId: "sc-1w8qzv8-6"
})(["align-items:center;display:flex;flex-direction:column;height:100%;min-width:70px;"]);
export var ModeIcon = styled.div.withConfig({
  displayName: "styled__ModeIcon",
  componentId: "sc-1w8qzv8-7"
})(["float:left;width:32px;height:32px;"]);
export var PrintableItinerary = styled.div.withConfig({
  displayName: "styled__PrintableItinerary",
  componentId: "sc-1w8qzv8-8"
})([""]);
