// @opentripplanner/itinerary-body/esm/demos/index.js
import coreUtils from "@opentripplanner/core-utils";
import PropTypes from "prop-types";
import React from "react";
import { action } from "@storybook/addon-actions";
import styled from "styled-components";
import { ExclamationTriangle } from "@styled-icons/fa-solid/ExclamationTriangle";
import ItineraryBody from "..";
import OtpRRTransitLegSubheader from "../otp-react-redux/transit-leg-subheader";
import * as ItineraryBodyClasses from "../styled";
export function CustomPlaceName(_ref) {
  var place = _ref.place;
  return "\uD83C\uDF89\u2728\uD83C\uDF8A ".concat(place.name, " \uD83C\uDF89\u2728\uD83C\uDF8A");
}
/**
 * Custom component, for illustration purposes only, for displaying the time and other info
 * of the given leg in the time column of the ItineraryBody -> PlaceRow component.
 */

export function CustomTimeColumnContent(_ref2) {
  var isDestination = _ref2.isDestination,
      leg = _ref2.leg,
      timeOptions = _ref2.timeOptions;
  var time = isDestination ? leg.endTime : leg.startTime;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("span", {
    style: {
      color: "#E60000"
    }
  }, time && coreUtils.time.formatTime(time, timeOptions))), /*#__PURE__*/React.createElement("div", {
    style: {
      fontSize: "80%",
      lineHeight: "1em"
    }
  }, /*#__PURE__*/React.createElement(ExclamationTriangle, {
    style: {
      height: "1em"
    }
  }), " Delayed xx\xA0min."));
}
CustomTimeColumnContent.propTypes = {
  isDestination: PropTypes.bool.isRequired,
  leg: coreUtils.types.legType.isRequired,
  timeOptions: coreUtils.types.timeOptionsType
};
CustomTimeColumnContent.defaultProps = {
  timeOptions: null
};
export function customToRouteAbbreviation(route) {
  if (route.toString().length < 3) {
    return route;
  }

  return undefined;
}
export function CustomTransitLegSummary(_ref3) {
  var leg = _ref3.leg,
      onClick = _ref3.onClick,
      stopsExpanded = _ref3.stopsExpanded;
  return (
    /*#__PURE__*/

    /* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */
    React.createElement("div", {
      onClick: onClick
    }, leg.duration && /*#__PURE__*/React.createElement("span", null, "Ride ", coreUtils.time.formatDuration(leg.duration)), leg.intermediateStops && /*#__PURE__*/React.createElement("span", null, " (".concat(leg.intermediateStops.length + 1, " stops)"), /*#__PURE__*/React.createElement(ItineraryBodyClasses.CaretToggle, {
      expanded: stopsExpanded
    })))
  );
}
CustomTransitLegSummary.propTypes = {
  leg: coreUtils.types.legType.isRequired,
  onClick: PropTypes.func.isRequired,
  stopsExpanded: PropTypes.bool.isRequired
};
export var StyledItineraryBody = styled(ItineraryBody).withConfig({
  displayName: "demos__StyledItineraryBody",
  componentId: "ss5q7m-0"
})(["", "{background-color:pink;}", "{color:#676767;font-size:14px;padding-right:4px;padding-top:1px;text-align:right;vertical-align:top;width:60px;}"], ItineraryBodyClasses.LegBody, ItineraryBodyClasses.TimeColumn);
export function WrappedOtpRRTransitLegSubheader(_ref4) {
  var languageConfig = _ref4.languageConfig,
      leg = _ref4.leg;
  return /*#__PURE__*/React.createElement(OtpRRTransitLegSubheader, {
    languageConfig: languageConfig,
    leg: leg,
    onStopClick: action("Transit Stop Click")
  });
}
WrappedOtpRRTransitLegSubheader.propTypes = {
  languageConfig: coreUtils.types.languageConfigType.isRequired,
  leg: coreUtils.types.legType.isRequired
};
