// @opentripplanner/stop-viewer-overlay/esm/default-stop-marker.js
import _extends from "@babel/runtime/helpers/extends";
// TODO: UNDO WHEN TYPESCRIPTING!

/* eslint-disable react/forbid-prop-types */
import PropTypes from "prop-types";
import React from "react";
import { Popup, CircleMarker } from "react-leaflet";
export default function DefaultStopMarker(_ref) {
  var leafletPath = _ref.leafletPath,
      radius = _ref.radius,
      stop = _ref.stop;
  return /*#__PURE__*/React.createElement(CircleMarker
  /* eslint-disable-next-line react/jsx-props-no-spreading */
  , _extends({}, leafletPath, {
    center: [stop.lat, stop.lon],
    key: stop.id,
    radius: radius
  }), /*#__PURE__*/React.createElement(Popup, null, /*#__PURE__*/React.createElement("div", null, stop.name)));
}
DefaultStopMarker.propTypes = {
  // leafletPath: coreUtils.types.leafletPathType,
  leafletPath: PropTypes.object,
  radius: PropTypes.number,
  stop: PropTypes.object.isRequired // stop: coreUtils.types.stopLayerStopType.isRequired

};
DefaultStopMarker.defaultProps = {
  leafletPath: {
    color: "#000",
    fillColor: "cyan",
    fillOpacity: 1,
    weight: 3
  },
  radius: 9
};
